geonode/harvesting/harvesters/base.py

```diff
@@ -195,6 +203,54 @@ def finalize_resource_update(
     ) -> ResourceBase:
         return geonode_resource
 
+    def should_copy_resource(
+            self,
+            harvestable_resource: "HarvestableResource",  # noqa
+    ) -> bool:
+        """Return True if the worker is able to copy the remote resource
+
+        The base implementation just returns False. Subclasses must re-implement this method
+        if they support copying remote resources onto the local GeoNode.
+
+        """
+
+        return False
+
+    def copy_resource(
+            self,
+            harvestable_resource: "HarvestableResource",  # noqa
+            harvested_resource_info: HarvestedResourceInfo,
+    ) -> typing.Optional[str]:
+        """Copy a remote resource's data to the local GeoNode.
+
+        The base implementation provides a generic copy using GeoNode's `storage_manager`.
+        Subclasses may need to re-implement this method if they require specialized behavior.
+
+        """
+
+        url = harvested_resource_info.resource_descriptor.distribution.original_format_url
+        result = None
+        if url is not None:
+            parsed_url = urllib.parse.urlparse(url)
+            harvester = models.Harvester.objects.get(pk=self.harvester_id)
+            name_fragments = (
+                f"{harvester.name}-{harvester.id}",
+                parsed_url.path.replace("/", "-")
+            )
+            target_name = "_".join(name_fragments)
+            response = requests.get(url)
+            if response.status_code == requests.codes.ok:
+                buffer = File(io.BytesIO(), name=target_name)
+                with buffer.open("wb") as fh:
+                    for chunk in response.iter_content(chunk_size=128):
+                        fh.write(chunk)
+                result = storage_manager.save(target_name, buffer)
+            else:
+                logger.error(f"Could not copy resource from {url}: {response.status_code} - {response.reason}")
+        else:
+            logger.warning("harvested resource info does not provide a URL for retrieving the resource, skipping...")
+        return result
```

Comment:
@afabiani @giohappy 

Can you review this and comment whether this is a fair usage of the `storage_manager`?

I'm basically storing the whole remote file in a temporary in-memory buffer and then calling `storage_manager.save()` with the buffer as an argument. I'm not sure this is how it is supposed to be used, as it feels weird having to have the whole file in-memory before passing all contents to storage manager (I can imagine this not being very performant for big files). Maybe I'm doing it wrong and I can write chunks iteratively using the storage manager directly?

Please advise.


Also, I'm not clear on how I will proceed after having stored the file locally via the storage manager. How do I now tell the `resource_manager`, which will get called to create/update the resource on the local GeoNode, that the file that is related to the `ResourceBase` being edited is in the local location?

Can you post some short pseudocode of how this works? Thanks

Reply:
I think I got it now. Working on an updated version.